### ISSUE TITLE
weird "magic" with no_headerless 0 (or macros, or linker?)

### ISSUE THREAD

**Issue by aniou** [mgcaret/of816#9]
Hi,
recently I made some observations about generated code - I'm rather pretty sure, that I found some kind of bug, but I'm not able to fully understood it nature.

Some background: some months ago I started port for C256 (thanks for referencing to my project). After putting whole thing into usable shape (at least I was able to transfer files by my xmodem version, included in fcode/xmodem.fs) I got break and few days ago I returned to next step - an ANSI terminal support. 

And then magic begins. Even very simple changes leads to "instabilities" - resets, hangs, etc. on real hw. Unfortunately official emulator still has problems with code and my own project it works fine (or I made incorrect tests), so I was forced to stick with real machine. 
I made all tests, to be sure that problems doesn't come from C256's construction and kernel (overlapping rom areas, clearing during reset routines etc). Nope. I made also multiple tests and checks about memory locations of stacks: os, return and parameters. Still nothing.

During test I realize, that instability appears when size platform-lib.s grows. So - thoughts about page crossings or something similar come to my mind. I replace all new code by multiple "NOP" and, after  several compilations, I got "a good" and "bad" (with one more NOP inside) code. Then I started comparing binaries (I made something like that before, during work on byte-to-byte exact conversion of LiaraForth to 64tass), so I was brave ;)

To make a long story shorter: I found some irregularities in code and I was pretty sure that I found bug in linker or assembler (I tested even latest version from repo, BTW) but today I found something that leads me to conclusion that clue is in macro expansion or ca65 or in linker itself: at this moment I'm not able to judge. From my POV looks like  that code compiled with no_headerless 0 was invalid and "good" and "bad" binaries was only a matter of luck.

As reminder: my version usess memory at 0x3a:0000

### Probably perfectly working code with no_headerless 1

Listing file from ```ca65 -l forth.lst``` with placeholder values manually replaced by data from binaries
As You can see there looks like perfectly valid code (and it works!)

```
"really good" with no_headerless 1
00 00 00 00   dword     XYZZY,"XYZZY",F_IMMED
85 58 59 5a 
5a 59 a0 
22 7b 01 3a   ENTER
7c 0e 3a 00   .dword FORTH
b9 01 3a 00   EXIT

00 00 00 00   hword     WLNUM,"WL#
83 57 4c 23   
20 
22 7b 01 3a   ENTER
c3 0e 3a 00   .dword dORDER
b4 21 3a 00   .dword SWAP
```

...and ```ca65 -v -v -v -v``` output looks like (linker debug, in all cases, shows proper addresses):
```
"really good" with no_headerless 1
  Literal: EA EA EA EA EA EA EA EA EA EA EA                  / my placeholder at end of platform-libs.s
           00 00 00 00                                       / dword macro
           85 58 59 5A 
           5A 59 A0 
           22  Expression (3):  SYM( SEC $0136 +)            / enter
               Expression (4):  SYM( SEC $0E37 +)            / dword forth
               Expression (4):  SYM( SEC $0175 +)  $0001 -   / exit

  Literal: 00 00 00 00                                       / hword (dword) macro WLNUM
           83 57 4C 23 
           20 
           22 Expression (3):  SYM( SEC $0136 +)             / enter
              Expression (4):  SYM( SEC $0E7E +)             / dword dorder
              Expression (4):  SYM( SEC $216F +)             / dword swap
```

### "good", but probably also invalid code with no_headerless 0

Take a look, please at hword WLNUM and ENTER macro result
```
00 00 00 00   dword     XYZZY,"XYZZY",F_IMMED
85 58 59 5A 
5A 59 A0 
22 7B 01 3A   ENTER
74 0E 3A 00   .dword FORTH
B9 01 3A 00   EXIT

80 20         hword     WLNUM,"WL#
22 7B 00 01   ENTER                             !!! (should be 22 7b 01 3a)?
3a a6 0e 3a   .dword dORDER
00 72 20 3a   .dword SWAP
...
```

and ca65 verbose:
```
  Literal: EA EA EA EA EA EA EA EA EA EA                      / my placeholder at end of platform-libs.s
           00 00 00 00                                        / dword macro
           85 58 59 5A
           5A 59 A0 
           22  Expression (3):  SYM( SEC $0136 +)             / enter
               Expression (4):  SYM( SEC $0E2F +)             / dword forth
               Expression (4):  SYM( SEC $0175 +)  $0001 -    / exit
  Literal: 80 20                                              / hword macro
           22  Expression (3):  SYM( SEC $0136 +)             / enter
               Expression (4):  SYM( SEC $0E61 +)             / dword dorder
               Expression (4):  SYM( SEC $202D +)             / dword swap

```

### "bad" and unusable code with no_headerless 0

The difference is in single NOP at and of platform-lib.s, so  single-byte differences are expected - but take a look at ENTER macro in WLNUM word - there is also byte swap (00 7b 01) in comparision to previous, shorter by one byte, code: (7b 00 01)  

```
00 00 00 00   dword     XYZZY,"XYZZY",F_IMMED
85 58 59 5A 
5A 59 A0 
22 7B 01 3A   ENTER
75 0E 3A 00   .dword FORTH
B9 01 3A 00   EXIT

80 20         hword     WLNUM,"WL#
22 00 7B 01   ENTER                   !!! as above + 00 7b 01 in place of 7b 00 01
3a a7 0e 3a   .dword dORDER
00 73 20 3a   .dword SWAP
```

compare that with copy from previous point - there is also a byte swap, that is a really mystery for me:
```
; from compilation with *lib.s code shorter by one byte
80 20         hword     WLNUM,"WL#
22 7B 00 01   ENTER                             !!! (should be 22 7b 01 3a)?
3a a6 0e 3a   .dword dORDER
00 72 20 3a   .dword SWAP
```

ca65 output section is the same:

```
  Literal: EA EA EA EA EA EA EA EA EA EA EA                   / my placeholder at end of platform-libs.s
           00 00 00 00 
           85 58 59 5A 
           5A 59 A0 
           22  Expression (3):  SYM( SEC $0136 +)             / enter 
               Expression (4):  SYM( SEC $0E30 +)             / dword forth
               Expression (4):  SYM( SEC $0175 +)  $0001 -    / exit
  Literal: 80 20 
           22  Expression (3):  SYM( SEC $0136 +)             / enter
               Expression (4):  SYM( SEC $0E62 +)             / dword dorder
               Expression (4):  SYM( SEC $202E +)             / dword swap
```

**Comment by aniou** [comment]
As side note: I used latest version of ./inc/macros.inc from Your repo.

**Comment by mgcaret** [comment]
How bizarre.

**Comment by mgcaret** [comment]
(also, sorry it took so long to answer, I tend to develop in random spurts depending on what is happening elsewhere in my life)

What version of ca65 are you using?  Right now my development machine is running "ca65 V2.18 - Git e0cb33d9"...  Did you build ca65 yourself?  If not, where did you get it?  If so, what C compiler and platform?

You'll have noted that I have a test suite (in /test) that I run with an emulator of my own that I created specifically for developing OF816 (GoSXB).  Current coverage is about 82% of the public (always have headers) words.  I had never run it with no_headerless = 1 until now.  The suite passes either way, but I have not tried making changes by adding NOPs to platform-lib.s.  I will play around with that.

I also have a TeamCity instance that compiles every commit automatically against a relatively recent version of ca65 (billing itself as "ca65 V2.18 - N/A").  I just compared the output binaries to those made by my local ca65, and they are the same. 

The output of the suite run against GoSXB just now is:
```
Executing suite: 7.3.1 Stack functions
Lines complete.
testing 7.3.1.1 Stack duplication
testing 7.3.1.2 Stack removal
testing 7.3.1.3 Stack rearrangement
testing 7.3.1.4 Return stack
testing 7.3.1.5 Stack depth
Errors = 0
Suite complete.
Executing suite: 7.3.2.1 Arithmetic - Single-precision integer arithmetic
Lines complete.
testing 7.3.2.1 Single-precision integer arithmetic
Errors = 0
Suite complete.
Executing suite: 7.3.2.2 Arithmetic - Bitwise logical operators
Lines complete.
testing 7.3.2.2 Bitwise logical operators
Errors = 0
Suite complete.
Executing suite: 7.3.2.3 Arithmetic - Double number arithmetic
Lines complete.
testing 7.3.2.3 Double number arithmetic
TESTING D+ with small integers
TESTING D+ with mid range integers
TESTING D+ with large double integers
TESTING D- with small integers
TESTING D- with mid-range integers
TESTING D- with large integers
Errors = 0
Suite complete.
Executing suite: 7.3.2.4 Arithmetic - Data type conversion
Lines complete.
testing 7.3.2.4 Data type conversion
Errors = 0
Suite complete.
Executing suite: 7.3.2.5 Arithmetic - Address arithmetic
Lines complete.
testing 7.3.2.5 Address arithmetic
Errors = 0
Suite complete.
Executing suite: 7.3.3 Memory control
Lines complete.
testing 7.3.3.1 Memory access
testing 7.3.3.2 Memory allocation
Errors = 0
Suite complete.
Executing suite: 7.3.4 Text input and output
Lines complete.
testing 7.3.4.1 Text input
testing 7.3.4.2 Console input
testing 7.3.4.3 ASCII constants
testing 7.3.4.4 Console output
testing 7.3.4.5 Output formatting
testing 7.3.4.6 Display pause
testing 7.3.4.7 String literals
testing 7.3.4.8 String manipulation
Errors = 0
Suite complete.
Executing suite: 7.3.5 Numeric input and output
Lines complete.
testing 7.3.5.1 Numeric-base control
testing 7.3.5.2 Numeric input
testing 7.3.5.3 Numeric output
testing 7.3.5.4 Numeric output primitives
Errors = 0
Suite complete.
Executing suite: 7.3.6 Comparison operators
Lines complete.
testing 7.3.6 Comparison operators
Errors = 0
Suite complete.
Executing suite: 7.3.7 Flag constants
Lines complete.
testing 7.3.7 Flag constants
Errors = 0
Suite complete.
Executing suite: 7.3.8 Control-flow commands
Lines complete.
testing 7.3.8.1 Conditional branches
testing 7.3.8.2 Case statement
testing 7.3.8.3 Conditional loops
testing 7.3.8.4 Counted loops
testing 7.3.8.5 Other control flow commands
testing 7.3.8.6 Error handling
Errors = 0
Suite complete.
Executing suite: 7.3.8-i Control-flow commands in interpretation state
Lines complete.
testing 7.3.8.1 Conditional branches - interpretation state
testing 7.3.8.2 Case statement - interpretation state
testing 7.3.8.3 Conditional loops - interpretation state
testing 7.3.8.4 Counted loops - interpretation state
testing 7.3.8.6 Error handling - interpretation state
Errors = 0
Suite complete.
Executing suite: 7.3.9 Forth dictionary
Lines complete.
testing 7.3.9.1 Defining words
testing 7.3.9.2 Dictionary commands
testing 7.3.9.2.1 Data space allocation
testing 7.3.9.2.2 Immediate words
testing 7.3.9.2.3 Dictionary search
testing 7.3.9.2.4 Miscellaneous dictionary
testing 7.3.9.2.4 Assembler
Errors = 0
Suite complete.
Executing suite: 7.5.1 Automatic stack display
Lines complete.
testing 7.5.1 Automatic stack display
Errors = 0
Suite complete.
Executing suite: 7.5.3 Dictionary
Lines complete.
testing 7.5.3 Dictionary
testing 7.5.3.1 Dictionary search
testing 7.5.3.2 Decompiler
Errors = 0
Suite complete.
Executing suite: 5.3.2 Forth FCode Functions
Lines complete.
testing 5.3.2 Forth FCode Functions
Errors = 0
Suite complete.
Executing suite: 5.3.3 FCode implementation functions
Lines complete.
testing 5.3.3 FCode implementation functions
testing 5.3.3.1 Defining new FCode functions
testing 5.3.3.3 Controlling values and defers
testing OF816 5.3.3-related checks
Errors = 0
Suite complete.
Executing suite: 5.3.4.3 Get local arguments
Lines complete.
testing 5.3.4.3 Get local arguments
Errors = 0
Suite complete.
Executing suite: 5.3.7.1 Peek/poke
Lines complete.
testing 5.3.7.1 Peek/poke
Errors = 0
Suite complete.
Executing suite: 5.3.7.2 Device-register access
Lines complete.
testing 5.3.7.1 Device-register access
Errors = 0
Suite complete.
Executing suite: 5.3.7.4 System-information
Lines complete.
testing 5.3.7.4 System information
Errors = 0
Suite complete.
Executing suite: 5.3.7.6 Start and end
Lines complete.
testing 5.3.7.6 Start and end
Errors = 0
Suite complete.
Executing suite: OF816 Words
Lines complete.
testing OF816 words
Errors = 0
Suite complete.
Tests complete, no errors.
```

Anyway, thanks for this report and I will play around with it.

**Comment by mgcaret** [comment]
I just tried adding 1 through 6 NOPs to platform-lib.s for GoSXB and could not replicate the results.

You should be able to compile the W65C816SXB port, it's the simplest of them and was my reference hardware until I developed GoSXB.  It's still one of two hardware platforms I test against, though not as often as the Neon816.

I am curious if you see the same effects by adding NOPs to its platform-lib.s... that might tell us whether your ca65/ld65 is being strange or not.  You might also try moving the FStartup segment to the end of the memory and see if that changes anything.

As it stands, I am at a loss.

**Comment by aniou** [comment]
Finally, I made tests with newest ca65 available (2.19) - and problem disappeared.
